File: app/src/main/java/com/zybooks/circuitmaker/MainActivity.java
package com.zybooks.circuitmaker;

import android.Manifest;
import android.content.ClipData;
import android.content.ClipDescription;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.net.Uri;
import android.os.Bundle;
import android.provider.MediaStore;
import android.util.Log;
import android.view.DragEvent;
import android.view.View;
import android.widget.Button;
import android.widget.ImageButton;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

import com.mihir.drawingcanvas.drawingView;

import java.io.File;
import java.io.FileOutputStream;

public class MainActivity extends AppCompatActivity {

    private static final int STORAGE_PERMISSION_CODE = 100;
    private View tools;
    private drawingView draw;
    private ConstraintLayout circuitView;
    private ImageButton undo, redo, stroke;
    private Button clear, save;
    private float buttonX;
    private float buttonY;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        if (ContextCompat.checkSelfPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE)
            != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE},
                    STORAGE_PERMISSION_CODE);
        }

//        draw = findViewById(R.id.circuit_view);
//        undo = findViewById(R.id.undo);
//        redo = findViewById(R.id.redo);
//        clear = findViewById(R.id.clean);
//        save = findViewById(R.id.save);
//
//        undo.setOnClickListener(v -> draw.undo());
//        redo.setOnClickListener(v -> draw.redo());
//        clear.setOnClickListener(v -> draw.clearDrawingBoard());
//
//        save.setOnClickListener(v -> draw.post(() -> {
//            Bitmap bitmap = getBitmapFromView(draw);
//            Uri imageUri = saveBitmapToStorage(bitmap);
//
//            if (imageUri != null) {
//                Log.e("draw", "Image saved at: " + imageUri);
//                Toast.makeText(this, "Image saved!", Toast.LENGTH_SHORT).show();
//                shareImage(imageUri);
//            } else {
//                Log.e("draw", "Saving failed");
//                Toast.makeText(this, "Failed to save image", Toast.LENGTH_SHORT).show();
//            }
//        }));

        tools = findViewById(R.id.circuit_tools);
        ImageButton button = findViewById(R.id.toolBox);
        Button placeholderButton = findViewById(R.id.placeholder_button);

        button.setOnClickListener(v -> {
            Log.d("BUTTONS", "clicked");
            int toolsVisible = tools.getVisibility();

            Log.d("X position when last dropped", String.valueOf(buttonX));
            Log.d("Y position when last dropped", String.valueOf(buttonY));

            Log.d("Width of tools container", String.valueOf(tools.getWidth()));
            Log.d("Height of tools container", String.valueOf(tools.getHeight()));

            if (toolsVisible == View.VISIBLE) {
                tools = findViewById(R.id.circuit_tools);
                tools.setVisibility(View.GONE);
                if (placeholderButton.getX() < tools.getWidth()) {
                    placeholderButton.setVisibility(View.GONE);
                } else {
                    placeholderButton.setX((float) (buttonX + (tools.getWidth() / 2.0) - 1));
                    placeholderButton.setY((float) (buttonY - (tools.getHeight() / 2.0) - 1));
                }
            } else {
                tools.setVisibility(View.VISIBLE);
                if (placeholderButton.getX() >= tools.getWidth()) {
                    placeholderButton.setX((float) (buttonX - (tools.getWidth() / 2.0) + 1));
                    placeholderButton.setY((float) (buttonY + (tools.getHeight() / 2.0) + 1));
                }
                placeholderButton.setVisibility(View.VISIBLE);
            }

            Log.d("Visibility of tools container", String.valueOf(tools.getVisibility()));
            Log.d("X position of placeholderButton", String.valueOf(placeholderButton.getX()));
            Log.d("Y position of placeholderButton", String.valueOf(placeholderButton.getY()));
        });

        circuitView = findViewById(R.id.main);

        onLongClickListener(circuitView, placeholderButton);

        // Set drag listener for the View
        circuitView.setOnDragListener(this::dragListener);
    }

//    private Bitmap getBitmapFromView(View view) {
//        Bitmap bitmap = Bitmap.createBitmap(view.getWidth(), view.getHeight(),
//                Bitmap.Config.ARGB_8888);
//        Canvas canvas = new Canvas(bitmap);
//        if (view.getBackground() != null) {
//            view.getBackground().draw(canvas);
//        } else {
//            canvas.drawColor(Color.WHITE);
//        }
//        view.draw(canvas);
//        return bitmap;
//    }
//
//    private Uri saveBitmapToStorage(Bitmap bitmap) {
//        String filename = "Drawing " + System.currentTimeMillis() + ".png";
//        Uri uri = null;
//
//        try {
//            File imagesDir = new File(getExternalFilesDir(null) + "/Pictures/DrawingApp");
//
//            if (!imagesDir.exists()) {
//                imagesDir.mkdirs();
//            }
//
//            File imageFile = new File(imagesDir, filename);
//            FileOutputStream outputStream = new FileOutputStream(imageFile);
//            bitmap.compress(Bitmap.CompressFormat.PNG, 100, outputStream);
//            outputStream.flush();
//            outputStream.close();
//
//            MediaStore.Images.Media.insertImage(getContentResolver(),
//                    imageFile.getAbsolutePath(), filename, null);
//            uri = Uri.fromFile(imageFile);
//        } catch (Exception e) {
//            Log.e("draw", "Error saving image: " + e.getMessage());
//            e.printStackTrace();
//        }
//
//        return uri;
//    }
//
//    private void shareImage(Uri uri) {
//        Intent shareIntent = new Intent(Intent.ACTION_SEND);
//        shareIntent.putExtra(Intent.EXTRA_STREAM, uri);
//        shareIntent.setType("image/png");
//        startActivity(Intent.createChooser(shareIntent, "Share via"));
//    }

    @NonNull
    private Button cloneButton(DragEvent e) {

        Log.d("Local state", String.valueOf(((Button) e.getLocalState()).getX()));
        Log.d("Tools x", String.valueOf(tools.getX()));

        if (tools.getWidth() > ((Button) e.getLocalState()).getX()) {

            Button cloneButton = new Button(MainActivity.this);
            Button originalButton = ((Button) e.getLocalState());

            cloneButton.setText(originalButton.getText());
            cloneButton.setTextColor(originalButton.getCurrentTextColor());
            cloneButton.setBackground(originalButton.getBackground());

            onLongClickListener(circuitView, cloneButton);

            return cloneButton;
        }
        return ((Button) e.getLocalState());
    }

    private void onLongClickListener(View v, Button button) {

        button.setOnLongClickListener( view -> {

            ClipData.Item item = new ClipData.Item((CharSequence) view.getTag());

            ClipData dragData = new ClipData(
                    (CharSequence) view.getTag(),
                    new String[] {ClipDescription.MIMETYPE_TEXT_PLAIN},
                    item);

            View.DragShadowBuilder myShadow = new View.DragShadowBuilder(button);

            view.startDragAndDrop(dragData,
                    myShadow,
                    view,
                    0);

            return true;
        });
    }

    @NonNull
    private Boolean dragListener(View v, DragEvent e) {
        // Handle each of the expected events
        switch(e.getAction()) {

            case DragEvent.ACTION_DRAG_STARTED:

                // Determine whether this View can accept
                if (e.getClipDescription().hasMimeType(ClipDescription.MIMETYPE_TEXT_PLAIN)) {

                    v.invalidate();

                    return true;
                }

                return false;

            case DragEvent.ACTION_DRAG_ENTERED:

                // ((ImageView) v).setColorFilter(Color.GREEN);

                v.invalidate();

                return true;

            case DragEvent.ACTION_DRAG_LOCATION:

                return true;

            case DragEvent.ACTION_DROP:

                ClipData.Item item = e.getClipData().getItemAt(0);

                CharSequence dragData = item.getText();

                Button cloneButton = cloneButton(e);
                if (cloneButton.getX() <= tools.getX()) {
                    circuitView.addView(cloneButton);
                }

                buttonX = e.getX();
                buttonY = e.getY();
                Log.d("Drop X position", String.valueOf(buttonX));
                Log.d("Drop Y position", String.valueOf(buttonY));

                cloneButton.setX(buttonX);
                cloneButton.setY(buttonY);

                Toast.makeText(this, "Dragged data is " + dragData, Toast.LENGTH_LONG).show();

                // ((ImageView) v).clearColorFilter();

                v.invalidate();

                return true;

            case DragEvent.ACTION_DRAG_ENDED:

                // ((ImageView) v).clearColorFilter();

                v.invalidate();

                if (e.getResult()) {
                    Toast.makeText(this, "The drop was handled.", Toast.LENGTH_LONG).show();

                } else {
                    Toast.makeText(this, "The drop didn't work.", Toast.LENGTH_LONG).show();
                }

                return true;

            default:
                Log.e("DragDrop test", "Unknown action type received by View.OnDragListener");
                break;
        }

        return false;
    }
}

File: app/src/main/java/com/zybooks/circuitmaker/MainActivity.kt
package com.zybooks.circuitmaker

import android.content.ClipData
import android.content.ClipDescription
import android.os.Bundle
import android.util.Log
import android.view.DragEvent
import android.view.View
import android.view.View.DragShadowBuilder
import android.widget.Button
import android.widget.ImageButton
import android.widget.Toast
import androidx.activity.enableEdgeToEdge
import androidx.appcompat.app.AppCompatActivity
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offest
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.tooling.preview.Preview
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat

class MainActivity : AppCompatActivity() {

    private var tools: View? = null

    private var circuitView: ConstraintLayout? = null

    private var buttonX = 0f
    private var buttonY = 0f
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()
        setContentView(R.layout.activity_main)
        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main)) { v, insets ->
            val systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom)
            insets
        }

        tools = findViewById(R.id.circuit_tools)
        val button: ImageButton? = findViewById(R.id.toolBox)
        val placeholderButton: Button? = findViewById(R.id.placeholder_button)

        button!!.setOnClickListener {
            Log.d("BUTTONS", "Clicked")
            val toolsVisible: Int? = tools!!.visibility

            Log.d("X position when last dropped", buttonX.toString())
            Log.d("Y position when last dropped", buttonY.toString())

            Log.d("Width of tools container", tools!!.width.toString())
            Log.d("Height of tools container", tools!!.height.toString())

            if (toolsVisible == View.VISIBLE) {
                tools!!.visibility = View.GONE
                if (placeholderButton!!.x < tools!!.width) {
                    placeholderButton.visibility = View.GONE
                } else {
                    placeholderButton.x = (buttonX + (tools!!.width / 2.0) - 1).toFloat()
                    placeholderButton.y = (buttonY + (tools!!.height / 2.0) - 1).toFloat()
                }
            } else {
                tools!!.visibility = View.VISIBLE
                placeholderButton?.x?.let {
                    if (it >= tools!!.width) {
                        placeholderButton.x = (buttonX - (tools!!.width / 2.0)).toFloat()
                        placeholderButton.y = (buttonY + (tools!!.height / 2.0)).toFloat()
                    }
                    placeholderButton.visibility = View.VISIBLE
                }
            }
        }
        circuitView = findViewById(R.id.main)
        onLongClickListener(placeholderButton)
        circuitView?.setOnDragListener(this::dragListener)
    }

    @Composable
    private fun DrawLine() {

    }

    private fun cloneButton(e: DragEvent): Button {
        Log.d("Local state", (e.localState as Button).x.toString())
        Log.d("Tools x", tools!!.x.toString())

        if (tools!!.width > (e.localState as Button).x) {
            val cloneButton = Button(this@MainActivity)
            val originalButton = (e.localState as Button)

            cloneButton.text = originalButton.getText()
            cloneButton.setTextColor(originalButton.currentTextColor)
            cloneButton.background = originalButton.background

            onLongClickListener(cloneButton)

            return cloneButton
        }
        return (e.localState as Button)
    }

    private fun onLongClickListener(button: Button?) {
        button?.setOnLongClickListener { view: View? ->
            val item = ClipData.Item(view!!.tag as CharSequence?)
            val dragData = ClipData(
                view.tag as CharSequence?,
                arrayOf(ClipDescription.MIMETYPE_TEXT_PLAIN),
                item
            )

            val myShadow = DragShadowBuilder(button)

            view.startDragAndDrop(
                dragData,
                myShadow,
                view,
                0
            )
            true
        }
    }

    private fun dragListener(v: View, e: DragEvent): Boolean {
        // Handle each of the expected events
        when (e.action) {
            DragEvent.ACTION_DRAG_STARTED -> {
                // Determine whether this View can accept
                if (e.clipDescription.hasMimeType(ClipDescription.MIMETYPE_TEXT_PLAIN)) {
                    v.invalidate()

                    return true
                }

                return false
            }

            DragEvent.ACTION_DRAG_ENTERED -> {
                // ((ImageView) v).setColorFilter(Color.GREEN);
                v.invalidate()

                return true
            }

            DragEvent.ACTION_DRAG_LOCATION -> return true

            DragEvent.ACTION_DROP -> {
                val item = e.clipData.getItemAt(0)

                val dragData = item.text

                val cloneButton = cloneButton(e)
                if (cloneButton.x <= tools!!.x) {
                    circuitView!!.addView(cloneButton)
                }

                buttonX = e.x
                buttonY = e.y
                Log.d("Drop X position", buttonX.toString())
                Log.d("Drop Y position", buttonY.toString())

                cloneButton.x = buttonX
                cloneButton.y = buttonY

                Toast.makeText(this, "Dragged data is $dragData", Toast.LENGTH_LONG).show()

                // ((ImageView) v).clearColorFilter();
                v.invalidate()

                return true
            }

            DragEvent.ACTION_DRAG_ENDED -> {
                // ((ImageView) v).clearColorFilter();
                v.invalidate()

                if (e.result) {
                    Toast.makeText(this, "The drop was handled.", Toast.LENGTH_LONG).show()
                } else {
                    Toast.makeText(this, "The drop didn't work.", Toast.LENGTH_LONG).show()
                }

                return true
            }

            else -> Log.e("DragDrop test", "Unknown action type received by View.OnDragListener")
        }

        return false
    }
}